### PR TITLE
Feat/react lifecycle

### PR DESCRIPTION
**Pull request by BiggA94**
I followed the Discussion in [here](https://github.com/microsoft/ReSub/issues/78).
I think, my Solution is not the best, because there is a Map, that contains the single instances of all the Components. Also there is a 'instanceId' in every State, so that I can get the instance via the State. 
But there should not be any big change in how ComponentBase works and how to use it (except for not using UNSAFE methods, because are not called when getDerivedStateFromProps is available)

This is a **BREAKING CHANGE**, due to the fact, that all the deprecated methods, and the methods, that start with UNSAFE_ are not called anymore.

What do you think?

### CODE REVIEW

**Comment by deregtd** [comment]
Thanks so much for taking a crack at this!  This is pretty much what I had envisioned trying out.  Still digging through it, but one quick question:

Out of curiosity, why did you end up going with the instance id and map instead of just straight up storing a "this" pointer into state?

**Comment by BiggA94** [comment]
> Out of curiosity, why did you end up going with the instance id and map instead of just straight up storing a "this" pointer into state?

I don't figured out my problem, but you can see my take on saving the instance directly in the store [here ](https://github.com/BiggA94/ReSub/blob/feat/pointer/src/ComponentBase.ts).

**Comment by ms-markda** [comment]
My main concern/paranoia is if React Fibers can abandon the component before it has mounted, and thus will it call unmount? If a component can be constructed but never mounted/unmount then keeping a dictionary of instances will leak.

React recommends not doing any work with side-effects in constructor or componentWillMount, although I cannot find a definitive source outlining these details :/

**Comment by berickson1** [comment]
I agree with Mark - however it's not fully clear how fiber will impact things but it's not any worse than the current state of the world, so I still think we should move forward with a change

A few additional comments -
I believe this changeset will result in calling buildState and setting up subscriptions twice on every render loop, which could be a perf hit. I added a test in #119 to ensure we don't regress this. I think this will involve some non-ideal logic to accomplish unfortunately
I think the typings on getDerivedStateFromProps could be improved

**Comment by BiggA94** [comment]
React [discourages us to save a Component inside the state](https://github.com/reduxjs/redux/issues/1248#issuecomment-172403969). And it doesn't work for me.
But I just woke up wiith the Idea of using a instance supplier Method inside the state.
This should fix the Concerns with the Memory Leak because of the Map.

I will check about the problem with calling buildState twice.

**Comment by ms-markda** [comment]
> This should fix the Concerns with the Memory Leak because of the Map.

The Map is the leak. Entries are only removed in unmount, but unmount is not guaranteed to be called unless the Component is mounted. Not a problem for now. Fiber can construct a Component instance and throw it away without ever mounting it. There is no "destructor" we can hook into. That is what makes this problem tricky 😄

**Comment by ms-markda** [comment]
However, this is the backbone of the full solution. Putting more hidden stuff in `state` is inevitable to avoid leaking. Keeping a Map of instance stuff will be useful to connect the `static` function to the instance. There are a few ways to put the pieces together that can be ironed out later.

This helps the current state of the world, aside from the double buildState issue, so indeed we should go forward with it.

**Comment by BiggA94** [comment]
I just learned, that property initializers of subclasses are called **after** the constructor of the superclass. That's why I had to move the buidLinitialState call from the constructor to the getDerviceStateFromProps.
This also fixes the problem of calling build state twice before mounting, but there seem to be to many calls later on. This should be fixed, but I think, that doesn't have to be part of this pull request.

File: src/ComponentBase.ts [hunk 1]
```diff
@@ -44,7 +44,13 @@ interface StoreSubscriptionInternal<P, S> extends StoreSubscription<P, S> {
     _subscriptionKey?: string;
 }
 
-export abstract class ComponentBase<P extends {}, S extends Dictionary<any>> extends React.Component<P, S> {
+interface InternalState {
+    _getInstance: () => ComponentBase<any, any>;
```
Comment (berickson1) [inline src/ComponentBase.ts]: I'd prefer to name this _resub_getInstance or something to indicate that it's the library doing this

File: test/AutoSubscribe.spec.tsx [hunk 2]
```diff
@@ -155,11 +174,10 @@ function makeComponent(props: SimpleProps): ReactWrapper<SimpleProps, SimpleStat
         storeDatas,
     } = Component.state();
 
-    // Internal check: state should have one change.
+    // Internal check: state should have two changes (including instanceId).
```
Comment (berickson1) [inline test/AutoSubscribe.spec.tsx]: I don't think this comment is true anymore

**Comment by berickson1** [comment]
I think we also need to update tslint rules with this change as well

**Comment by berickson1** [comment]
I pointed your commit against a new "dev" branch, so we can sort out all the test/perf/changes for 2.x without affecting 1.x in master

**Comment by a-tarasyuk** [comment]
@berickson1 which changes need to add to linting rules?

**Comment by berickson1** [comment]
We need to update the ReSub rules for super calls and ensure it works with static methods

**Comment by BiggA94** [comment]
I think the most important is, that components that use getDerivedStateFromProps must return at least the return value from super.getDerivedStateFromProps (But they can extend that object).
I think this could be a difficult rule.. 
If this is too difficult, then it should be sufficient imho to test, if the static function of the super class is at leat called.

The other thing is, that react only shows a warning, that the old methods are not called anymore..
I think, every Component that is a Component Base **must not** declare this functions.
So I think this should also be a linting rule.

**Pull request merged**